Chinsky's cardiopulmonary medical bonanza.

comma opened the pull request:
Will add changelog later.
-Pulse affecting blood pressure.
-Inapro moving pulse towards normal rather than always lowering it.
-2fast pulse having chance to damage heart.
-Oxygenation / pain heart stoppage happening with prob() rather than 100% to make CPR more rewarding
-Shock levels start rising at lower damage (30 vs 50) and can reach higher than equal to shock value, so you can actually reach that 120 level needed for heartstop without passing out.
-CPR uses lung breathe processing instead of straight up oxyloss healing,

comma committed:
Opoids pharmacokinetics derpery and moar injectors

Makes painkillers process way faster in blood (not in stomach). Ups the requiered dosage too.
In effect, when you eat a pill you get current behaviour in terms of how long it works for, but it ramps up for ~minute.
If oyu inject, it stays way less, but it ramps up twice as fast.
Adds more autoinjector types, adds them to vendomats.

comma committed:
Pulse and hearttsopping changes

It modifies effective one, slow pulse gives you 90%, high 110%, 2high 125%.
2fast and thready pulse now have chance to damage heart, so it's not always a good idea to use them to boost blood levels.
Inapro no longer just lowers the pulse, it instead moves it one step towards normal pulse.
Oxygenation now affects pulse - increased pulse at first, thready pulse at bad values.
Changed heart stoppage to be bit more 'fuzzier', only hard stop is low blood circulation now. Traumatic shock and low oxygen (checking for thready pulse now) now use prob to see if they should stop the heart. Should make CPR slightly less meaningless, since heart would be actually to work for a bit when restarted.

comma committed:
Shock level changes

Now can start at 30 traumatic shock instead of 50.
Maximum shock stage is more than shock value, so that 120 (heartstop) as achievable. Currently to achieve it you need traumatic shock big enough to KO you, which lowers your shock, and you're stuck in a cycle of pass out get up without ever reaching it.
Makes KO shock modifier bit higher since it was way too stronk.

comma committed:
Makes CPR use lungs instead of directly adjust oxyloss

It gets breath of CPRer, then uses victim's lungs to process it. In perfect conditions it's same result - 5 oxyloss healed, BUT, this also considers things like NOT HAVING AIR WHILE YOU ARE DOING CPR (because you can totally do that currently), lung damage, poison in the air, etc.
Also buffs chance to fracture ribs a bit cause apparently it never fukn happens.

comma commented:
todo: consider 2fast damage cause it would usually get to that with shock/bloodloss and trigger spiral with damaging heart.

Techhead0 commented on code/modules/organs/internal/heart.dm:
```diff
@@ -26,30 +26,50 @@
+	var/should_stop = owner.get_blood_circulation() < BLOOD_VOLUME_SURVIVE //cardiovascular shock, not enough liquid to pump
+	should_stop = should_stop || prob(max(0, owner.getBrainLoss() - owner.maxHealth * 0.75)) //brain failing to work heart properly
+	should_stop = should_stop || (prob(10) && owner.shock_stage >= 120) //traumatic shock
+	should_stop = should_stop || (prob(60) && pulse == PULSE_THREADY) //erratic heart patterns, usually caused by oxyloss
```
Feels like a really high chance here. Makes it more likely to stop than not, every tick.

comma replied:
Kinda intended, since it's a big deal and means they're suffocating pretty badly (gonna adjust threshold for thisw pulse to SURVIVE from BAD). Can be countered by inapro too.

PurpleMartinJCK commented on code/modules/mob/living/carbon/human/human_attackhand.dm:
```diff
@@ -92,14 +92,20 @@
-					adjustOxyLoss(-(min(getOxyLoss(), 5)))
-					updatehealth()
-					to_chat(src, "<span class='notice'>You feel a breath of fresh air enter your lungs. It feels good.</span>")
+					var/species_organ = species.breathing_organ
```
You should use need_breathe() here https://github.com/Baystation12/Baystation12/blob/5c07e22eebf0dc0cb46cb124ef1b4da2ece75246/code/modules/mob/living/carbon/human/human.dm#L1571

PurpleMartinJCK commented on code/modules/mob/living/carbon/human/human_attackhand.dm:
```diff
@@ -92,14 +92,20 @@
 				H.visible_message("<span class='notice'>\The [H] performs CPR on \the [src]!</span>")
-				if(prob(3))
+				if(prob(5))
 					var/obj/item/organ/external/chest = get_organ(BP_CHEST)
```
You're splitting the breath from the heart part pretty well here. You should adjust it so that they can still try to start the heart without lungs https://github.com/comma/Baystation12/blob/531822b3b5839481f26f336e5f1f8988e37c51ec/code/modules/mob/living/carbon/human/human_attackhand.dm#L79 as well as the whole mask bit.

comma replied:
That's quite a split on CPR, but might as well I guess

PurpleMartinJCK replied:
Yeah lemme know if it's too much for you right now as while it'd be nice, it's probably not needed for this PR.

comma replied:
Naw, just moved all those checks to before breathing part.

comma committed:
Some additional changes

Fuzzifies heartstop a bit more, even with not enough blood you might get a beat or two if lucky.
Moves thready pulse to lower oxygenation level, BAD was too high for heart-stop effect.
Lowers 2FAST pulse damage, 0.5 damage with 1% chance should be good enough to discourage casual heart overdrives while not being solid enough to present actual danger when people are damaged.

comma committed:
Changelog
